internal/expr/scan.go

```diff
@@ -0,0 +1,196 @@
+package expr
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+func (re *ResultExpr) One(args ...any) error {
+	if ok, err := re.Next(); err != nil {
+		return err
+	} else if !ok {
+		return fmt.Errorf("cannot return one row: no results")
+	}
+	err := re.Decode(args...)
+	if err != nil {
+		return err
+	}
+	re.Close()
+	return nil
+}
+
+// getTypes returns the types in the order they appear in the query
+func getTypes(outs []field) []reflect.Type {
+	isDup := make(map[reflect.Type]bool)
+	ts := []reflect.Type{}
+	for _, out := range outs {
+		if t := out.structType; !isDup[t] {
+			isDup[t] = true
+			ts = append(ts, t)
+		}
+	}
+	return ts
+}
+
+// All returns a slice containing all rows returned in the query. Each row is
+// a slice of any types that contains all structs mentioned in the output expressions
+// of the query.
+func (re *ResultExpr) All() ([][]any, error) {
+	var s [][]any
+
+	ts := getTypes(re.outputs)
+
+	for {
+		if ok, err := re.Next(); err != nil {
+			return [][]any{}, err
+		} else if !ok {
+			break
+		}
+
+		rs := []any{}
+		for _, t := range ts {
+			rp := reflect.New(t)
+			// We need to unwrap the struct inside the interface{}.
+			r := rp.Elem()
+			err := re.decodeValue(r)
+			if err != nil {
+				return [][]any{}, err
+			}
+			rs = append(rs, r.Interface())
+		}
+
+		s = append(s, rs)
+	}
+
+	re.Close()
+	return s, nil
+}
+
+// Next calls sql.rows.Next() and checks if it has thrown an error.
+func (re *ResultExpr) Next() (bool, error) {
+	ok := re.rows.Next()
+	if !ok && re.rows.Err() != nil {
+		return false, re.rows.Err()
+	}
+	return ok, nil
+}
+
+// Decode copies the columns in the current row into the fields of the structs in dests.
+// dests can contain any subset of the structs mentioned in the query.
+func (re *ResultExpr) Decode(dests ...any) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot decode expression: %s", err)
+		}
+	}()
+
+	for _, dest := range dests {
+		if dest == nil {
+			return fmt.Errorf("need valid sturct, got nil")
+		}
+
+		v := reflect.ValueOf(dest)
+		if v.Kind() != reflect.Pointer {
+			return fmt.Errorf("need pointer to struct, got non-pointer")
+		}
+		v = reflect.Indirect(v)
+
+		if v.Kind() != reflect.Struct {
+			return fmt.Errorf("need struct, got %s", v.Kind())
+		}
+
+		if err := re.decodeValue(v); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (re *ResultExpr) scan() ([]any, error) {
+	cols, err := re.rows.Columns()
+	if err != nil {
+		return []any{}, fmt.Errorf("cannot scan: %s", err)
+	}
+
+	// rows.Scan takes a list of pointers to where it should scan the values
+	// from the result row.
+	// We call rows.Scan(ptrs) and the results are stored in vals.
+	vals := make([]any, len(cols))
+	ptrs := make([]any, len(cols))
+	for i, _ := range cols {
+		ptrs[i] = &vals[i]
+	}
+	re.rows.Scan(ptrs...)
+
+	// offset is the distace between column X in the row and the position of the
```

Comment:
I thought maybe we could rephrase this in a different way, but again, ignore if you think it's not clearer :)

sqlair columns are named as _sqlair_X, where X = 0, 1, 2, .... 
offset is the difference between X and the list index where _sqlair_X is located in the list of columns of the result row.

Reply:
Changed